src/utils/theme.js

```diff
@@ -72,29 +144,68 @@ function activateTheme() {
         `textHoverColor: #fff; textColor: #fff; backgroundColor: ${actionColor}; backgroundHoverColor: ${actionHoverColor}`
       );
   }
+}
+
+const themeChangedListeners = new Map();
+function registerThemeChangedListener(listener) {
+  window.APP.store.addEventListener("themechanged", listener);
+  const [_darkModeQuery, removeDarkModeListener] = registerDarkModeQuery(listener);
 
-  window.dispatchEvent(new CustomEvent("update_theme"));
+  themeChangedListeners.set(listener, {
+    removeListener: () => {
+      window.APP.store.removeEventListener("themechanged", listener);
+      removeDarkModeListener();
+    }
+  });
+}
+
+function removeThemeChangedListener(listener) {
+  themeChangedListeners.get(listener).removeListener();
+  themeChangedListeners.delete(listener);
 }
 
-const onStoreChanged = (function() {
-  let themeId;
-  return function onStoreChanged() {
-    const newThemeId = window.APP?.store?.state?.preferences?.theme;
-    if (themeId !== newThemeId) {
-      themeId = newThemeId;
-      activateTheme();
+// window.APP.store may not be available when onThemeChanged is called, so we
+// stash listeners until waitForDOMContentLoaded.
+const stashedThemeChangedListeners = new Set();
+function onThemeChanged(listener) {
+  const storeIsAvailable = !!(window.APP?.store);
+  if (storeIsAvailable) {
+    registerThemeChangedListener(listener);
+  } else {
+    stashedThemeChangedListeners.add(listener);
+  }
+
+  return () => {
+    if (storeIsAvailable) {
```

Comment:
I don't think `storeIsAvailable` will be re-computed here, so a listener that started in the stash and then registered via `registerStashedThemeChangedListeners` may not be cleaned up.

We should either recompute `!!(window.APP?.store)` or handle both possible states:
```js
stashedThemeChangedListeners.delete(listener);
if (themeChangedListeners.has(listener)) {
  themeChangedListeners.get(listener).removeListener();
  themeChangedListeners.delete(listener);
}
```

Reply:
Ah, yeah, wasn't sure what to do here. Handling both states is a good idea. Thanks.